File: 0542-01-matrix/0542-01-matrix.java
class Solution {
    class Node{
        int r;
        int c;
        int dis;

        Node(int r, int c, int dis){
            this.r = r;
            this.c = c;
            this.dis = dis;
        }
    }

    public int[][] updateMatrix(int[][] mat) {
        int m = mat.length; 
        int n = mat[0].length;

        boolean[][] vis = new boolean[m][n];
        int[][] res = new int[m][n];

        Queue<Node> q = new LinkedList<>();

        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(mat[i][j] == 0){
                    q.add(new Node(i, j, 0));
                    vis[i][j] = true;
                }
            }
        }

        int[][] directions = {{0,1}, {1,0}, {-1,0}, {0,-1}};

        while(!q.isEmpty()){
            Node curr = q.poll();

            for(int[] dir: directions){
                int nx = curr.r + dir[0];
                int ny = curr.c + dir[1];
                if(nx>=0 && nx<m && ny>=0 && ny<n && vis[nx][ny]==false && res[nx][ny]<=curr.dis+1){
                    vis[nx][ny] = true;
                    res[nx][ny] = curr.dis+1;
                    q.add(new Node(nx, ny, curr.dis+1));
                }   
            }

        }

        return res;
    }
}

File: 0542-01-matrix/0542-01-matrix.py
from collections import deque

class Solution(object):
    def updateMatrix(self, mat):
        """
        :type mat: List[List[int]]
        :rtype: List[List[int]]
        """
        m, n = len(mat), len(mat[0])
        res = [[float('inf')] * n for _ in range(m)]
        q = deque()

        for i in range(m):
            for j in range(n):
                if mat[i][j] == 0:
                    res[i][j] = 0
                    q.append((i, j))

        directions = [(0,1), (1,0), (-1,0), (0,-1)]
        while q:
            x, y = q.popleft()
            for dx, dy in directions:
                nx, ny = x + dx, y + dy
                if 0 <= nx < m and 0 <= ny < n:
                    if res[nx][ny] > res[x][y] + 1:
                        res[nx][ny] = res[x][y] + 1
                        q.append((nx, ny))

        return res
